File: frontend/src/types/enterprise.ts
// 企业相关类型定义

/**
 * 企业状态
 */
export type EnterpriseStatus = 'active' | 'pending' | 'inactive';

/**
 * 成员角色
 */
export type EnterpriseRole = 'owner' | 'admin' | 'member' | 'viewer';

/**
 * 成员状态
 */
export type MemberStatus = 'active' | 'pending' | 'inactive';

/**
 * 企业成员用户信息
 */
export interface MemberUser {
  /** 用户ID */
  id: string;
  /** 用户邮箱 */
  email: string;
  /** 用户姓名 */
  name: string;
  /** 用户头像 */
  avatar?: string | null;
}

/**
 * 企业成员
 * 与 types/index.ts 中的 EnterpriseMember 保持兼容
 */
export interface EnterpriseMember {
  /** 成员记录ID */
  id: string;
  /** 企业ID - 两种命名风格兼容 */
  enterpriseId?: string;
  enterprise_id?: string;
  /** 用户ID - 两种命名风格兼容 */
  userId?: string;
  user_id?: string;
  /** 成员角色 */
  role: EnterpriseRole;
  /** 成员状态 - 可选 */
  status?: MemberStatus;
  /** 加入时间 - 两种命名风格兼容 */
  joinedAt?: string;
  joined_at?: string;
  /** 用户信息 - 可选 */
  user?: MemberUser;
  /** 兼容 index.ts 的字段 */
  username?: string;
  email?: string;
  userEmail?: string;
  full_name?: string;
  avatar_url?: string;
}

/**
 * 企业设置
 */
export interface EnterpriseSettings {
  /** 成员加入需要审批 */
  requireApproval?: boolean;
  /** 允许公开查看 */
  allowPublicView?: boolean;
  /** 默认成员角色 */
  defaultMemberRole?: EnterpriseRole;
  /** 通知设置 */
  notificationSettings?: {
    emailEnabled?: boolean;
    newMemberAlert?: boolean;
    roleChangeAlert?: boolean;
  };
}

/**
 * 企业信息
 * 统一的 Enterprise 类型，合并了 enterprise.ts 和 index.ts 的定义
 */
export interface Enterprise {
  /** 企业ID */
  id: string;
  /** 企业名称 */
  name: string;
  /** 企业描述 - 可选 */
  description?: string;
  /** 企业地址 - 可选 */
  address?: string;
  /** 联系邮箱 - 两种命名风格兼容 */
  contactEmail?: string;
  contact_email?: string;
  /** 联系电话 - 可选 */
  contactPhone?: string;
  /** 官方网站 - 可选 */
  website?: string;
  /** 所属行业 - 可选 */
  industry?: string;
  /** 企业规模 - 可选 */
  scale?: string;
  /** 企业状态 - 两种表示方式兼容 */
  status?: EnterpriseStatus;
  is_active?: boolean;
  /** 兼容 isActive 属性 */
  isActive?: boolean;
  /** Logo URL - 可选 */
  logo_url?: string;
  /** 钱包地址 - 可选 */
  wallet_address?: string;
  /** 是否已验证 - 可选 */
  is_verified?: boolean;
  /** 成员数量 - 可选 */
  member_count?: number;
  /** 创建时间 - 两种命名风格兼容 */
  createdAt?: string;
  created_at?: string;
  /** 更新时间 - 两种命名风格兼容 */
  updatedAt?: string;
  updated_at?: string;
  /** 成员列表 - 仅在企业详情中包含 */
  members?: EnterpriseMember[];
  /** 企业设置 - 仅在企业详情中包含 */
  settings?: EnterpriseSettings;
}

/**
 * 企业详情 - 包含成员列表
 * 与 Enterprise 类型兼容，确保 members 字段存在
 */
export interface EnterpriseDetail extends Enterprise {
  members: EnterpriseMember[];
}

/**
 * 企业列表响应
 */
export interface EnterpriseListResponse {
  items: Enterprise[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

/**
 * 创建企业请求
 * 合并了两种命名风格的字段
 */
export interface EnterpriseCreateRequest {
  /** 企业名称 - 必填 */
  name: string;
  /** 企业描述 - 可选 */
  description?: string;
  /** 企业地址 - 可选 */
  address?: string;
  /** 联系邮箱 - 两种命名风格兼容 */
  contactEmail?: string;
  contact_email?: string;
  /** 联系电话 - 可选 */
  contactPhone?: string;
  /** 官方网站 - 可选 */
  website?: string;
  /** 所属行业 - 可选 */
  industry?: string;
  /** 企业规模 - 可选 */
  scale?: string;
  /** Logo URL - 可选 */
  logo_url?: string;
}

/**
 * 更新企业请求
 * 所有字段都是可选的
 */
export interface EnterpriseUpdateRequest {
  name?: string;
  description?: string;
  address?: string;
  contactEmail?: string;
  contact_email?: string;
  contactPhone?: string;
  website?: string;
  industry?: string;
  scale?: string;
  logo_url?: string;
}

/**
 * 邀请成员请求
 * 支持两种命名风格
 */
export interface InviteMemberRequest {
  /** 被邀请人邮箱 - 用于通过邮箱邀请 */
  email?: string;
  /** 被邀请人用户ID - 用于直接邀请已知用户 */
  user_id?: string;
  /** 角色 */
  role: EnterpriseRole | MemberRole;
}

/**
 * 更新成员角色请求
 */
export interface UpdateMemberRoleRequest {
  role: EnterpriseRole | MemberRole;
}

// 重新导出 index.ts 中的 MemberRole 类型以保持兼容
export type MemberRole = import('./index').MemberRole;

File: backend/app/schemas/enterprise.py
"""Enterprise-related request and response schemas."""

from datetime import datetime
import re
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.models.enterprise import MemberRole


EMAIL_REGEX = r"^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$"


class EnterpriseCreateRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    name: str = Field(..., min_length=2, max_length=100, description="Enterprise name")
    description: Optional[str] = Field(None, max_length=1000, description="Enterprise description")
    logo_url: Optional[str] = Field(None, max_length=500, description="Enterprise logo URL")
    website: Optional[str] = Field(None, max_length=255, description="Enterprise website")
    contact_email: Optional[str] = Field(
        None,
        max_length=255,
        description="Contact email",
        alias="contactEmail",
    )
    address: Optional[str] = Field(None, max_length=500, description="Enterprise address")

    @field_validator("name")
    @classmethod
    def validate_name(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("Enterprise name cannot be empty")
        return value

    @field_validator("website")
    @classmethod
    def validate_website(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return value
        value = value.strip()
        if not value:
            return None
        if not re.match(r"^https?://", value):
            raise ValueError("Website must start with http:// or https://")
        return value

    @field_validator("contact_email")
    @classmethod
    def validate_contact_email(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return value
        value = value.strip().lower()
        if not value:
            return None
        if not re.match(EMAIL_REGEX, value):
            raise ValueError("Invalid email format")
        return value


class EnterpriseUpdateRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    name: Optional[str] = Field(None, min_length=2, max_length=100, description="Enterprise name")
    description: Optional[str] = Field(None, max_length=1000, description="Enterprise description")
    logo_url: Optional[str] = Field(None, max_length=500, description="Enterprise logo URL")
    website: Optional[str] = Field(None, max_length=255, description="Enterprise website")
    contact_email: Optional[str] = Field(
        None,
        max_length=255,
        description="Contact email",
        alias="contactEmail",
    )
    address: Optional[str] = Field(None, max_length=500, description="Enterprise address")

    @field_validator("name")
    @classmethod
    def validate_name(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return value
        value = value.strip()
        if not value:
            raise ValueError("Enterprise name cannot be empty")
        return value

    @field_validator("website")
    @classmethod
    def validate_website(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return value
        value = value.strip()
        if not value:
            return None
        if not re.match(r"^https?://", value):
            raise ValueError("Website must start with http:// or https://")
        return value

    @field_validator("contact_email")
    @classmethod
    def validate_contact_email(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return value
        value = value.strip().lower()
        if not value:
            return None
        if not re.match(EMAIL_REGEX, value):
            raise ValueError("Invalid email format")
        return value


class MemberResponse(BaseModel):
    id: str = Field(..., description="Member relation ID")
    user_id: str = Field(..., description="User ID")
    username: str = Field(..., description="Username")
    email: str = Field(..., description="User email")
    full_name: Optional[str] = Field(None, description="Full name")
    avatar_url: Optional[str] = Field(None, description="Avatar URL")
    role: MemberRole = Field(..., description="Member role")
    joined_at: datetime = Field(..., description="Joined at")

    class Config:
        from_attributes = True


class EnterpriseResponse(BaseModel):
    id: str = Field(..., description="Enterprise ID")
    name: str = Field(..., description="Enterprise name")
    description: Optional[str] = Field(None, description="Enterprise description")
    logo_url: Optional[str] = Field(None, description="Enterprise logo URL")
    website: Optional[str] = Field(None, description="Enterprise website")
    contact_email: Optional[str] = Field(None, description="Contact email")
    address: Optional[str] = Field(None, description="Enterprise address")
    wallet_address: Optional[str] = Field(None, description="Wallet address")
    is_active: bool = Field(..., description="Active flag")
    is_verified: bool = Field(..., description="Verified flag")
    created_at: datetime = Field(..., description="Created at")
    updated_at: datetime = Field(..., description="Updated at")
    member_count: int = Field(default=0, description="Member count")

    class Config:
        from_attributes = True


class EnterpriseDetailResponse(EnterpriseResponse):
    members: List[MemberResponse] = Field(default_factory=list, description="Members")


class EnterpriseListResponse(BaseModel):
    items: List[EnterpriseResponse] = Field(..., description="Enterprise list")
    total: int = Field(..., description="Total count")
    page: int = Field(..., description="Page number")
    page_size: int = Field(..., description="Page size")
    pages: int = Field(..., description="Total pages")


class BoundWalletEnterpriseResponse(BaseModel):
    enterprise_id: str = Field(..., description="Enterprise ID")
    enterprise_name: str = Field(..., description="Enterprise name")
    wallet_address: str = Field(..., description="Bound wallet address")
    is_active: bool = Field(..., description="Active flag")
    is_verified: bool = Field(..., description="Verified flag")
    member_count: int = Field(default=0, description="Member count")


class InviteMemberRequest(BaseModel):
    user_id: Optional[str] = Field(None, description="Invited user ID")
    email: Optional[str] = Field(None, max_length=255, description="Invited user email")
    role: MemberRole = Field(default=MemberRole.MEMBER, description="Member role")

    @model_validator(mode="after")
    def validate_invitee(self) -> "InviteMemberRequest":
        if not self.user_id and not self.email:
            raise ValueError("Either user_id or email is required")
        return self

    @field_validator("email")
    @classmethod
    def validate_email(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return value
        value = value.strip().lower()
        if not value:
            return None
        if not re.match(EMAIL_REGEX, value):
            raise ValueError("Invalid email format")
        return value

    @field_validator("role")
    @classmethod
    def validate_role(cls, value: MemberRole) -> MemberRole:
        if value == MemberRole.OWNER:
            raise ValueError("Cannot invite a member as owner")
        return value


class UpdateMemberRoleRequest(BaseModel):
    role: MemberRole = Field(..., description="New member role")

    @field_validator("role")
    @classmethod
    def validate_role(cls, value: MemberRole) -> MemberRole:
        if value == MemberRole.OWNER:
            raise ValueError("Cannot promote a member to owner")
        return value


class WalletBindChallengeRequest(BaseModel):
    wallet_address: str = Field(
        ...,
        min_length=42,
        max_length=42,
        description="Ethereum wallet address",
    )

    @field_validator("wallet_address")
    @classmethod
    def validate_wallet_address(cls, value: str) -> str:
        if not re.match(r"^0x[a-fA-F0-9]{40}$", value):
            raise ValueError("Invalid Ethereum wallet address")
        return value.lower()


class WalletBindChallengeResponse(BaseModel):
    challenge_token: str = Field(..., description="One-time bind challenge token")
    wallet_address: str = Field(..., description="Ethereum wallet address")
    message: str = Field(..., description="Message to be signed")
    expires_at: datetime = Field(..., description="Challenge expiration time")


class BindWalletRequest(BaseModel):
    wallet_address: str = Field(
        ...,
        min_length=42,
        max_length=42,
        description="Ethereum wallet address",
    )
    signature: str = Field(..., description="Signature payload")
    challenge_token: str = Field(..., min_length=16, description="One-time bind challenge token")

    @field_validator("wallet_address")
    @classmethod
    def validate_wallet_address(cls, value: str) -> str:
        if not re.match(r"^0x[a-fA-F0-9]{40}$", value):
            raise ValueError("Invalid Ethereum wallet address")
        return value.lower()
